searchlib/src/tests/attribute/searchable/attributeblueprint_test.cpp

```diff
@@ -210,5 +228,96 @@ TEST("requireThatFastSearchLocationTermsWork") {
+const BlueprintType*
+as_type(const Blueprint& blueprint)
+{
+    const auto* result = dynamic_cast<const BlueprintType*>(&blueprint);
+    ASSERT_TRUE(result != nullptr);
+    return result;
+}
+
+class NearestNeighborFixture {
+public:
+    MyAttributeManager mgr;
+    vespalib::string attr_name;
+    AttributeContext attr_ctx;
+    FakeRequestContext request_ctx;
+    AttributeBlueprintFactory source;
+
+    NearestNeighborFixture(AttributeVector::SP attr)
+        : mgr(attr),
+          attr_name(attr->getName()),
+          attr_ctx(mgr),
+          request_ctx(&attr_ctx),
+          source()
+    {
+    }
+    void set_query_tensor(const TensorSpec& tensor_spec) {
+        request_ctx.set_query_tensor("query_tensor", tensor_spec);
+    }
+    Blueprint::UP create_blueprint() {
+        query::NearestNeighborTerm term("query_tensor", attr_name, 0, Weight(0), 7);
+        return source.createBlueprint(request_ctx, FieldSpec(attr_name, 0, 0), term);
+    }
+};
+
+TEST_F("nearest neighbor blueprint is created by attribute blueprint factory",
+       NearestNeighborFixture(make_tensor_attribute(field, "tensor(x[2])")))
+{
+    TensorSpec dense_x_2 = TensorSpec("tensor(x[2])").add({{"x", 0}}, 3).add({{"x", 1}}, 5);
+    f.set_query_tensor(dense_x_2);
+
+    auto result = f.create_blueprint();
+    const auto* nearest = as_type<NearestNeighborBlueprint>(*result);
+    EXPECT_EQUAL("tensor(x[2])", nearest->get_attribute_tensor().getTensorType().to_spec());
+    EXPECT_EQUAL(dense_x_2, DefaultTensorEngine::ref().to_spec(nearest->get_query_tensor()));
+    EXPECT_EQUAL(7u, nearest->get_target_num_hits());
+}
+
+void
+expect_empty_blueprint(AttributeVector::SP attr, const TensorSpec& query_tensor, bool insert_query_tensor = true)
+{
+    NearestNeighborFixture f(attr);
+    if (insert_query_tensor) {
+        f.set_query_tensor(query_tensor);
+    }
+    auto result = f.create_blueprint();
+    as_type<EmptyBlueprint>(*result);
+}
+
+void
+expect_empty_blueprint(AttributeVector::SP attr)
+{
+    expect_empty_blueprint(std::move(attr), TensorSpec("double"), false);
+}
+
+TEST("empty blueprint is created when nearest neighbor term is invalid")
+{
+    TensorSpec sparse_x = TensorSpec("tensor(x{})").add({{"x", 0}}, 3);
+    TensorSpec dense_y_2 = TensorSpec("tensor(y[2])").add({{"y", 0}}, 3).add({{"y", 1}}, 5);
+    expect_empty_blueprint(make_int_attribute(field)); // attribute is not a tensor
+    expect_empty_blueprint(make_tensor_attribute(field, "tensor(x{})")); // attribute is not a dense tensor
+    expect_empty_blueprint(make_tensor_attribute(field, "tensor(x[2],y[2])")); // tensor type is not of order 1
+    expect_empty_blueprint(make_tensor_attribute(field, "tensor(x[2])")); // query tensor not found
+    expect_empty_blueprint(make_tensor_attribute(field, "tensor(x[2])"), sparse_x); // query tensor is not dense
+    expect_empty_blueprint(make_tensor_attribute(field, "tensor(x[2])"), dense_y_2); // tensor types are not equal
```

Comment:
should also test x[2] vs x[3] (not same size)
and float x[2] vs double x[2] -> ok

Reply:
Fixed